### ISSUE TITLE
Вопрос по авторизации по токену

### ISSUE THREAD

**Issue by ivan-koptev** [amocrm/amocrm-api-php#267]
Не совсем понял, есть ли возможность использовать API как автоматизированный способ аутентификации с моим аккаунтом amoCRM? Что я имею в виду: клиент заполняет форму, и мы с помощью API передаем информацию из формы в amoCRM (создаем контакт + сделку). Сейчас для получения токена, мы редиректимся на страницу, где выбираем аккаунт, а потом опять редиректимся на страницу, которая указана уже в настройках интеграции. Есть ли способ избежать эти шаги и создать прямую связку между сайтом и аккаунтом CRM?

**Comment by eugene-borovov** [comment]
Да, такой способ есть. Можно исключить клиентские редиректы путём прямой передачи авторизационного кода в методы API, а далее токены будут сами обновляться.

В примерах показан путь через редирект, но можно взять в настройках интеграции авторизационный код и пропустить ветку редиректов. https://github.com/amocrm/amocrm-api-php/blob/master/examples/get_token.php#L44

**Comment by artem-palamarchuk96** [comment]
@eugene-borovov а refresh_token сам не обновляется и получается интеграция падает раз в какоето время и надо вручную зайти и опять сделать ре-авторизацию чтобы обновить refresh_token, как это можно обойти чтобы интеграция никогда не падала b refresh_token сам обновлялся?

**Comment by eugene-borovov** [comment]
@artem-palamarchuk96 если вы внимательно посмотрите на пример, который я указал, то там настраивается сохранение токенов. В других примерах get_token.php всегда включён. Внутри методов библиотеки происходит автоматический перезапрос access token с помощью refresh token. После удачного обновления получается новая пара токенов и сохраняется способом определённым в get_token.php. Таким образом, если всё настроенно ноомально, refresh_token всегда свежий при условии постоянных запросов API.

**Comment by artem-palamarchuk96** [comment]
@eugene-borovov наверное тут ключевой момент "постоянных запросов", просто сайт где в моем случае это используется, обращение к API может быть раз в несколько месяцев, и я так понимаю что в этом и есть фишка, что токен просто не обновляется, а когда надо обновить он уже заекспайрился

**Comment by eugene-borovov** [comment]
В daily cron что-нибудь типа https://github.com/amocrm/amocrm-api-php/blob/master/examples/account_actions.php и на ошибку email админу. Токен всегда будет в актуальном состоянии.

**Comment by artem-palamarchuk96** [comment]
@eugene-borovov звучит как план, спасибо за помощь!
